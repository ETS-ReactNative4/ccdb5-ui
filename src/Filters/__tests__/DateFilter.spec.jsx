import ReduxDateFilter, { DateFilter, mapDispatchToProps } from '../DateFilter'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme';
import thunk from 'redux-thunk'

function setupEnzyme(initialProps={}) {
  const props = Object.assign({
    changeDateRange: jest.fn(),
    fieldName: 'date_received',
    from: '',
    maximumDate: '',
    minimumDate: '',
    through: '',
    title: 'Date CFPB Received the complaint'
  }, initialProps)

  const target = shallow(<DateFilter {...props} />);

  return {
    props,
    target
  }
}

function setupSnapshot(query={}) {
  const middlewares = [thunk]
  const mockStore = configureMockStore(middlewares)
  const store = mockStore({
    query
  })

  return renderer.create(
    <Provider store={store}>
      <ReduxDateFilter fieldName="date_received"
                       minimumDate="2015-01-01"
                       maximumDate="2015-12-31"
                       title="Date CFPB Received the complaint"
      />
    </Provider>
  )
}

describe('component::DateFilter', () => {
  describe('snapshot', () => {
    it('supports no dates in the Redux Store', () => {
      const target = setupSnapshot()
      const tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('shows errors', () => {
      const target = setupSnapshot({
        date_received_min: new Date(2016, 0, 1),
        date_received_max: new Date(2000, 0, 1)
      })
      const tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('date entry', () => {
    it('triggers an update when a valid from date is entered', () => {
      const { target, props } = setupEnzyme()
      const input = target.find('input[aria-describedby="input-error_message-from"]')
      input.simulate('change', { target: { value: '2015-06-07' }})
      const actual = props.changeDateRange.mock.calls[0]
      expect(actual[0]).toEqual(expect.any(Date))
      expect(actual[0].getFullYear()).toEqual(2015)
      expect(actual[0].getMonth()).toEqual(6-1)
      expect(actual[1]).toEqual(null)
    })

    it('triggers an update when a valid through date is entered', () => {
      const { target, props } = setupEnzyme()
      const input = target.find('input[aria-describedby="input-error_message-through"]')
      input.simulate('change', { target: { value: '2015-06-07' }})
      const actual = props.changeDateRange.mock.calls[0]
      expect(actual[1]).toEqual(expect.any(Date))
      expect(actual[1].getFullYear()).toEqual(2015)
      expect(actual[1].getMonth()).toEqual(6-1)
      expect(actual[0]).toEqual(null)
    })

    it('does not trigger an update when an invalid date is entered', () => {
      const { target, props } = setupEnzyme()
      const input = target.find('input[aria-describedby="input-error_message-from"]')
      input.simulate('change', { target: { value: '9999-99-9' }})
      expect(props.changeDateRange).not.toHaveBeenCalled()
      expect(target.state('messages').from).toEqual("'9999-99-9' is not a valid date.")
    })
  })

  describe('componentWillReceiveProps', () => {
    it('validates the new props', () => {
      const {target} = setupEnzyme()
      target.instance()._validate = jest.fn(() => ({}))
      target.setProps({from: '2015-01-01', to: '2015-12-31'})
      expect(target.instance()._validate).toHaveBeenCalled()
    })    

    it('does not trigger a new update', () => {
      const {target, props} = setupEnzyme()
      target.setProps({from: '2015-01-01', to: '2015-12-31'})
      expect(props.changeDateRange).not.toHaveBeenCalled()
    })    
  })

  describe('mapDispatchToProps', () => {
    it('hooks into changeDateRange', () => {
      const dispatch = jest.fn()
      const props = {fieldName: 'qaz'}
      mapDispatchToProps(dispatch, props).changeDateRange('foo', 'bar')
      expect(dispatch.mock.calls.length).toEqual(1)
    })
  })
})
