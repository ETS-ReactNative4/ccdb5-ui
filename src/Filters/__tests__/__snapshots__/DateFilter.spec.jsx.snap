exports[`component::DateFilter snapshot shows errors 1`] = `
<section
  className="o-expandable aggregation date-filter">
  <div
    className="o-expandable_header o-expandable_target">
    <h5
      className="o-expandable_header-left o-expandable_label">
      Date CFPB Received the complaint
    </h5>
    <span
      className="o-expandable_header-right o-expandable_link">
      <button
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}>
        Hide
        <span
          className="cf-icon cf-icon-minus-round" />
      </button>
    </span>
  </div>
  <p />
  <div
    className="layout-row">
    <div
      className="flex-all">
      <label
        className="a-label a-label__heading body-copy">
        From
      </label>
      <input
        aria-describedby="input-error_message-from"
        className="a-text-input__error"
        max="2015-12-31"
        min="2015-01-01"
        onChange={[Function]}
        type="date"
        value="2016-01-01" />
    </div>
    <div
      className="flex-all">
      <label
        className="a-label a-label__heading body-copy">
        Through
      </label>
      <input
        aria-describedby="input-error_message-through"
        className="a-text-input__error"
        max="2015-12-31"
        min="2015-01-01"
        onChange={[Function]}
        type="date"
        value="2000-01-01" />
    </div>
    <ul
      className="messages">
      <li
        className="a-error-message"
        id="input-error_message-from"
        role="alert">
        <span
          aria-hidden="true"
          className="cf-icon cf-icon-delete-round" />
        \'2016-01-01\' must be less than 2015-12-31
      </li>
      <li
        className="a-error-message"
        id="input-error_message-through"
        role="alert">
        <span
          aria-hidden="true"
          className="cf-icon cf-icon-delete-round" />
        \'2000-01-01\' must be greater than 2015-01-01
      </li>
      <li
        className="a-error-message"
        id="input-error_message-ordered"
        role="alert">
        <span
          aria-hidden="true"
          className="cf-icon cf-icon-delete-round" />
        \'From\' must be less than \'Through\'
      </li>
    </ul>
  </div>
</section>
`;

exports[`component::DateFilter snapshot supports no dates in the Redux Store 1`] = `
<section
  className="o-expandable aggregation date-filter">
  <div
    className="o-expandable_header o-expandable_target">
    <h5
      className="o-expandable_header-left o-expandable_label">
      Date CFPB Received the complaint
    </h5>
    <span
      className="o-expandable_header-right o-expandable_link">
      <button
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}>
        Hide
        <span
          className="cf-icon cf-icon-minus-round" />
      </button>
    </span>
  </div>
  <p />
  <div
    className="layout-row">
    <div
      className="flex-all">
      <label
        className="a-label a-label__heading body-copy">
        From
      </label>
      <input
        aria-describedby="input-error_message-from"
        className=""
        max="2015-12-31"
        min="2015-01-01"
        onChange={[Function]}
        type="date"
        value="" />
    </div>
    <div
      className="flex-all">
      <label
        className="a-label a-label__heading body-copy">
        Through
      </label>
      <input
        aria-describedby="input-error_message-through"
        className=""
        max="2015-12-31"
        min="2015-01-01"
        onChange={[Function]}
        type="date"
        value="" />
    </div>
  </div>
</section>
`;
